addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1122,6 +1125,62 @@ func (r *ClusterBootstrapReconciler) updateValuesFromSecret(cluster *clusterapiv
 	return newSecret, nil
 }
 
+// cloneEmbeddedLocalObjectRef attempts to clone the embedded local object references from provider's namespace to cluster's
+// namespace. An example of embedded local object reference is the secret reference under CPIConfig.
+func (r *ClusterBootstrapReconciler) cloneEmbeddedLocalObjectRef(cluster *clusterapiv1beta1.Cluster, provider *unstructured.Unstructured) error {
+	groupKindNamesMap := util.ExtractTypedLocalObjectRef(provider.UnstructuredContent(), constants.LocalObjectRefSuffix)
+	if len(groupKindNamesMap) == 0 {
+		return nil
+	}
+
+	providerGVK := provider.GroupVersionKind()
+	r.Log.Info(fmt.Sprintf("cloning the embedded local object references within provider: %s with name: %s from"+
+		" %s namespace to %s namespace", provider.GroupVersionKind().String(), provider.GetName(), provider.GetNamespace(), cluster.Namespace))
+	for groupKind, resourceNames := range groupKindNamesMap {
+		gvr, err := r.getGVR(groupKind)
+		if err != nil {
+			// error has been logged within getGVR()
+			return err
+		}
+		for _, resourceName := range resourceNames {
+			r.Log.Info(fmt.Sprintf("cloning the GVR %s with name %s from %s namespace to %s namespace",
+				gvr.String(), resourceName, provider.GetNamespace(), cluster.Namespace))
+			fetchedObj, err := r.dynamicClient.Resource(*gvr).Namespace(provider.GetNamespace()).Get(r.context, resourceName, metav1.GetOptions{})
+			if err != nil {
+				r.Log.Error(err, fmt.Sprintf("unable to get provider: %s with name: %s under namespace: %s",
+					providerGVK.String(), provider.GetName(), provider.GetNamespace()))
+				return err
+			}
+
+			copiedObj := fetchedObj.DeepCopy()
+			unstructured.RemoveNestedField(copiedObj.Object, "metadata", "resourceVersion")
+			copiedObj.SetNamespace(cluster.Namespace)
+			_, err = r.dynamicClient.Resource(*gvr).Namespace(cluster.Namespace).Create(r.context, copiedObj, metav1.CreateOptions{})
+			if err != nil {
+				if apierrors.IsAlreadyExists(err) {
+					var jsonData []byte
```

Comment:
Could we add log here to explicitly state that?
```
Resource %s already exist in target namespace %s, only patch the ownerRef and labels
```

Reply:
yes, it is a good idea.

Reply:
I guess I am missing where ownerRef and labels are being patched.

Reply:
I guess I am missing where ownerRef and labels are being patched.

Reply:
It would be nice if I could learn a little bit more on the requirements of `ownerRef and labels` for embedded object.

Take the following as an example, and have few questions:
```
apiVersion: cpi.tanzu.vmware.com/v1alpha1
kind: VSphereCPIConfig
metadata:
  name: test-cluster-cpi
  namespace: tkg-system
spec:
  vsphereCPI:
    mode: vsphereCPI
    region: test-region
    zone: test-zone
    insecureFlag: true
    vmInternalNetwork: internal-net
    vmExternalNetwork: external-net
    vmExcludeInternalNetworkSubnetCidr: 192.168.3.0/24
    vmExcludeExternalNetworkSubnetCidr: 22.22.3.0/24
    vSphereCredentialLocalObjRef:
      kind: Secret
      name: cpi-vsphere-credential
```
- Who is supposed to be the owner of secret `cpi-vsphere-credential` ?
  - VSphereCPIConfig ?
  - CAPI Cluster ?
- What labels do we have to add ?
  - `tkg.tanzu.vmware.com/package-name` ? I guess we do not want to add this label since this secret is not the data value secret for a package. It is only meaningful within the scope of VSphereCPIConfig.
  - `tkg.tanzu.vmware.com/cluster-name` ?  I do not think we want to add this, it will be enqueued for reconciliation by the event handler we set for secrets.

Reply:
added ownerrefs.

Reply:
added ownerrefs.